fairlearn/metrics/_annotated_metric_function.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) Fairlearn contributors.
+# Licensed under the MIT License.
+
+from typing import Callable, Dict, List, Optional
+
+import logging
+import pandas as pd
+
+logger = logging.getLogger(__name__)
+
+_DEFAULT_NAME = "metric"
+
+_METRIC_FUNCTION_NONE = "Found 'None' instead of metric function"
+_METRIC_FUNCTION_NOT_CALLABLE = "Object passed as metric function not callable"
+
+
+class AnnotatedMetricFunction:
+    """Wrapper for functions, to give them args and kwargs."""
+
+    def __init__(
+        self,
+        *,
+        func: Callable,
+        name: Optional[str],
+        postional_argument_names: List[str] = None,
+        kw_argument_mapping: Dict[str, str] = None
+    ):
+        if func is None:
+            raise ValueError(_METRIC_FUNCTION_NONE)
+        if not callable(func):
+            raise ValueError(_METRIC_FUNCTION_NOT_CALLABLE)
+        self.func = func
+
+        if name is None:
+            if hasattr(func, "__name__"):
+                self.name = func.__name__
+            else:
+                logger.warning("Supplied 'func' had no __name__ attribute")
+                self.name = _DEFAULT_NAME
+        else:
+            self.name = name
+
+        self.func = func
+        self.postional_argument_names = ["y_true", "y_pred"]
+        if postional_argument_names is not None:
+            self.postional_argument_names = postional_argument_names
+        self.kw_argument_mapping = dict()
+        if kw_argument_mapping is not None:
+            self.kw_argument_mapping = kw_argument_mapping
+
+    def invoke(self, df: pd.DataFrame, split_columns: Dict[str, List[str]]):
+        """Invoke the wrapped function on the supplied DataFrame.
+
+        The function extracts its arguments from the supplied DataFrame :code:`df`.
+        Columns listed in :code:`self.postional_argument_names` are supplied
+        positionally, while those
+        in :code:`self.kw_argument_mapping` are supplied as keyword arguments.
+
+        There are two subtleties.
+        Firstly, the keyword arguments expected by the function might not
+        have the same names as the columns in the supplied DataFrame.
+        This is the reason :code:`self.kw_argument_mapping` is a dictionary, rather than
+        just a list of column names.
+
+        The second issue is coping with when users have passed in a 2D array as
```

Comment:
I think we should be able to avoid this subtlety (see my comment below re. `_accrete_array`).

Reply:
I'm a little concerned about all the type changes, but this does allow me to ditch `_accrete_array()` which is a plus.